driver-core/src/main/com/mongodb/client/model/densify/DensifySpecification.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.client.model.densify;
+
+import com.mongodb.annotations.Evolving;
+import com.mongodb.client.model.Aggregates;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+/**
+ * A specification of how to densify a sequence of documents.
+ *
+ * @see Aggregates#densify(DensifySpecification)
+ * @mongodb.server.release 5.1
+ * @since 4.7
+ */
+@Evolving
+public interface DensifySpecification extends PartitionByFieldDensifySpecificationBase, Bson {
```

Comment:
The reason I did not call it `Densify`, which I like more than `DensifySpecification`, is that objects of this type do not represent a complete pipeline stage. The stage is created by `Aggregates.densify(DensifySpecification)`, and I wanted to avoid confusion that may arise by seeing `Densify` and assuming it represents a stage.

However, if with my answer you still think it is better to name this type `Densify`, I will gladly do so.

P.S. I am not sure if "densify" is made up, because both [Google](https://www.google.com/search?q=densify&ei=oLGrYoDcCMS-kPIPmdeiwA8&ved=0ahUKEwiArOe4hbP4AhVEH0QIHZmrCPgQ4dUDCA4&uact=5&oq=densify&gs_lcp=Cgdnd3Mtd2l6EAMyEQguELEDEIMBEMcBENEDEJECMgUIABCRAjIFCAAQkQIyBQgAEIAEMgUIABCABDIFCAAQgAQyCwguEIAEEMcBEK8BMgUIABCABDIFCAAQgAQyBQgAEIAEOgcIABBHELADSgQIQRgASgQIRhgAULUEWLUEYJsIaAJwAXgAgAFqiAFqkgEDMC4xmAEAoAEByAEIwAEB&sclient=gws-wiz) and dictionaries ([here](https://www.merriam-webster.com/dictionary/densify), [here](https://www.collinsdictionary.com/dictionary/english/densify)) seem to consider it a real word.

Reply:
> builders for those stages called Filters and Projections

They are called `Filter`**s** and `Projection`**s** because those classes do not define a type used by either driver or a user, instead, they are simply compartments for factory methods. The same is true for `Windows` and `WindowedComputations`. Notice how the actual types they build are called `Window` and `WindowedComputation` respectively.

The `$search` query building API went further: why having a separate class that is just a holder for factory methods, when those factory methods may be contained by the interface corresponding to the type which instances are constructed by those factory methods? (the ability to define `static` methods in interfaces was added to Java because it [_"makes it easier for you to organize helper methods in your libraries; you can keep static methods specific to an interface in the same interface rather than in a separate class"_](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html#static)).

Given the above, `DensifySpecification` may be renamed to `Densify` or `Densification`, but `Densifications` (plural) does not appear to be the right way to name this type.

> I think I prefer Densifications

Is the proposed name `Densification` instead of `Densify` intended to reduce the probability of user thinking that an instance of `Densification` is a complete `$densify` stage? If yes, then using this name may help in this case, but we may not always come of with similar names for other future stages. For example, doing it for `$out` or `bucket` does not seem possible. This is why I used `DensifySpecification` - this naming approach will fork for any stage. However, the approach of calling it `Densify` will also work for for any stage.

Reply:
Resolving because refactoring made the question irrelevant.